src/validation.cpp

```diff
@@ -1669,6 +1748,41 @@ class WarningBitsConditionChecker : public AbstractThresholdConditionChecker
 // Protected by cs_main
 static ThresholdConditionCache warningcache[VERSIONBITS_NUM_BITS];
 
+static unsigned int GetBlockScriptFlags(const CBlockIndex* pindex, const CChainParams& chainparams) {
```

Comment:
This function can take Consensus::Params directly instead of CChainParams
Can we pass ```ThresholdConditionCache& cache``` to this and move it to src/versionbits ?
I know not all flags are activated with bip9, but it seems the best place to me.
Doing so would prevent you from calling IsWitnessEnabled() and AssertLockHeld(cs_main), but do you need to?
You can do the latter from the caller like we do with VersionBitsState